File: tests/src/test_nparray.cpp
#include "test_nparray.h"

void bind_test_is_f_contiguous(py::module &m) {
    m.def(
        "is_f_contiguous",
        [](py::array_t<double> & arr) {return carma::is_f_contiguous(arr);},
        "Test is F contiguous"
    );
}

void bind_test_is_c_contiguous(py::module &m) {
    m.def(
        "is_c_contiguous",
        [](py::array_t<double> & arr) {return carma::is_c_contiguous(arr);},
        "Test is C contiguous"
    );
}

void bind_test_is_writable(py::module &m) {
    m.def(
        "is_writable",
        [](py::array_t<double> & arr) {return carma::is_writable(arr);},
        "Test is writable"
    );
}

void bind_test_is_owndata(py::module &m) {
    m.def(
        "is_owndata",
        [](py::array_t<double> & arr) {return carma::is_owndata(arr);},
        "Test is owndata"
    );
}

void bind_test_is_aligned(py::module &m) {
    m.def(
        "is_aligned",
        [](py::array_t<double> & arr) {return carma::is_aligned(arr);},
        "Test is aligned"
    );
}

void bind_test_set_not_owndata(py::module &m) {
    m.def(
        "set_not_owndata",
        [](py::array_t<double> & arr) {carma::set_not_owndata(arr);},
        "Test is set_not_owndata"
    );
}

void bind_test_set_not_writeable(py::module &m) {
    m.def(
        "set_not_writeable",
        [](py::array_t<double> & arr) {carma::set_not_writeable(arr);},
        "Test is set_not_writeable"
    );
}

File: tests/test_nparray.py
"""Test nparray.h."""
import numpy as np

import libs.test_carma as carma


def test_is_f_contiguous():
    """Test is_f_contiguous."""
    m = 'F order array should be F contiguous'
    sample = np.ones((10, 2), dtype=np.float64, order='F')
    assert carma.is_f_contiguous(sample) is True, m

    m = 'C order array should not be F contiguous'
    sample = np.ones((10, 2), dtype=np.float64, order='C')
    assert carma.is_f_contiguous(sample) is False, m


def test_is_c_contiguous():
    """Test is_c_contiguous."""
    m = 'C order array should be C contiguous'
    sample = np.ones((10, 2), dtype=np.float64, order='C')
    assert carma.is_c_contiguous(sample) is True, m

    m = 'F order array should not be C contiguous'
    sample = np.ones((10, 2), dtype=np.float64, order='F')
    assert carma.is_c_contiguous(sample) is False, m


def test_is_writeable():
    """Test is_writable."""
    m = 'Array should be writable'
    sample = np.ones((10, 2), dtype=np.float64, order='F')
    assert carma.is_writable(sample) is True, m

    m = 'Array should not be writable'
    sample.setflags(write=0)
    assert carma.is_writable(sample) is False, m


def test_is_owndata():
    """Test is_writable."""
    m = 'Array should own the data'
    sample = np.ones((10, 2), dtype=np.float64, order='F')
    assert carma.is_owndata(sample) == sample.flags['OWNDATA'], m

    m = 'Array should not own the data'
    view = sample.reshape(20, 1)
    assert carma.is_owndata(view) == view.flags['OWNDATA'], m


def test_is_aligned():
    """Test is_aligned."""
    m = 'Array should be aligned'
    sample = np.arange(200, dtype=np.float64)
    assert carma.is_aligned(sample) == sample.flags['ALIGNED'], m

    m = 'Array should not be aligned'
    alt = np.frombuffer(sample.data, offset=2, count=100, dtype=np.float64)
    alt.shape = 10, 10
    assert carma.is_aligned(alt) == alt.flags['ALIGNED'], m
